azurelinuxagent/common/logcollector.py

```diff
@@ -0,0 +1,302 @@
+
+
+class LogCollector(object):
+
+    _TRUNCATED_FILE_PREFIX = "truncated_"
+
+    def __init__(self, manifest_file_path):
+        self.manifest_file_path = manifest_file_path
+        self.must_collect_files = self._expand_must_collect_files()
+
+    @staticmethod
+    def run_shell_command(command, stdout=subprocess.PIPE, output=False):
+        def format_command(cmd):
+            return " ".join(cmd) if isinstance(cmd, list) else command
+
+        def _encode_command_output(output):
+            return ustr(output, encoding='utf-8', errors="backslashreplace")
+
+        try:
+            process = subprocess.Popen(command, stdin=subprocess.PIPE, stdout=stdout, stderr=subprocess.PIPE, shell=False)
+            stdout, stderr = process.communicate()
+            return_code = process.returncode
+        except Exception as e:
+            error_msg = u"Command [{0}] raised unexpected exception: [{1}]".format(format_command(command), ustr(e))
+            LogCollector._log_to_results_file(error_msg)
+            return
+
+        if return_code != 0:
+            encoded_stdout = _encode_command_output(stdout)
+            encoded_stderr = _encode_command_output(stderr)
+            error_msg = "Command: [{0}], return code: [{1}], stdout: [{2}] stderr: [{3}]".format(format_command(command),
+                                                                                                 return_code,
+                                                                                                 encoded_stdout,
+                                                                                                 encoded_stderr)
+            LogCollector._log_to_results_file(error_msg)
+            return
+
+        if output:
+            msg = "Output of command [{0}]:\n{1}".format(format_command(command), _encode_command_output(stdout))
+            LogCollector._log_to_results_file(msg)
+
+    @staticmethod
+    def _expand_must_collect_files():
+        # Match the regexes from the MUST_COLLECT_FILES list to existing file paths on disk.
+        manifest = []
+        for path in _MUST_COLLECT_FILES:
+            manifest.extend(sorted(glob.glob(path)))
+
+        return manifest
+
+    def _read_manifest_file(self):
+        return read_file(self.manifest_file_path).splitlines()
+
+    @staticmethod
+    def _log_to_results_file(entry):
+        if isinstance(entry, list):
+            append_items_to_file(_OUTPUT_RESULTS_FILE_PATH, entry)
+        else:
+            append_file(_OUTPUT_RESULTS_FILE_PATH, entry + "\n")
+
+    @staticmethod
+    def _process_ll_command(folder):
+        LogCollector.run_shell_command(["ls", "-alF", folder], output=True)
+
+    @staticmethod
+    def _process_echo_command(message):
+        LogCollector._log_to_results_file(message)
+
+    @staticmethod
+    def _process_copy_command(path):
+        file_paths = glob.glob(path)
+        LogCollector._log_to_results_file(file_paths)
+        return file_paths
```

Comment:
`copy` comes from the IID command in the manifest, e.g. `copy,/var/log/waagent.log` where the intended usage is to collect that file. In my implementation, that means adding that file to the file list and logging that addition to the results file, since I collect all the files in the end, after finalizing the file list. I've named all these helper processing functions to match the commands from the manifest (copy, echo, ll).